lib/libxdp/xsk.c

```diff
@@ -1093,6 +1217,27 @@ int xsk_umem__delete(struct xsk_umem *umem)
 	return 0;
 }
 
+static void xsk_release_xdp_prog(struct xsk_socket *xsk)
+{
+	struct xsk_ctx *ctx = xsk->ctx;
+	int value;
+
+	if (!xsk->ctx->refcnt_map_fd)
+		goto ret;
+
+	value = xsk_decr_prog_refcnt(ctx->refcnt_map_fd);
+	if (value < 0)
+		pr_warn("Error occurred when decrementing xsk XDP prog refcount: %d, please detach program yourself\n",
+			value);
+	if (value)
+		goto ret;
```

Comment:
It's a bit subtle that this if-statement also catches the `< 0` case above, so maybe change this to:

```C
/* Detach if we were the last user */
if (value == 0)
  xdp_program__detach(...)
```